app/Http/Middleware/VersionSwitcher.php

```diff
@@ -27,28 +26,17 @@ public function handle($request, $next)
 
         $route = $request->route();
 
-        $request->setRouteResolver(function () use ($version, $route) {
-
-            return array_map(function ($routeComponent) use ($version) {
-                // Only handle routes that has a controller defined.
-                if ($uses = $routeComponent['uses'] ?? null) {
-                    if (preg_match('/^([^@]+)@(.*)$/', $uses, $m)) {
-                        list(, $classFrom, $method) = $m;
-
-                        // Squeeze in a \v[version]\ before the controller class name.
-                        $replacement = sprintf('$1v%d\\\\$2', $version);
-                        $classTo = preg_replace('/(.*\\\\)([\w]+)$/', $replacement, $classFrom);
-
-                        // If we have a controller candidate use that.
-                        if (method_exists($classTo, $method)) {
-                            $routeComponent['uses'] = "$classTo@$method";
-                        }
-                    }
+        foreach ($route as $routeComponent) {
+            // Only handle routes that has a controller defined.
+            if ($uses = $routeComponent['uses'] ?? null) {
+                [$controllerFrom, $method] = explode('@', $uses);
+                $replacement = sprintf('$1v%d\\\\$2', $version);
+                $controllerTo = preg_replace('/(.*\\\\)([\w]+)$/', $replacement, $controllerFrom);
+                if (method_exists($controllerTo, $method)) {
+                    app()->alias($controllerTo, $controllerFrom);
```

Comment:
But you're still aliasing the real class to `ListController`? Rather than all that string manipulation (who understands that regex anyway?):

```php

            if ($uses = $routeComponent['uses'] ?? null) {
                [$controllerFrom, $method] = explode('@', $uses);
                if ($controllerFrom == 'list_controller') {
                  if (class_exists($controllerTo)) {
                      app()->alias($controllerTo, 'list_controller');
                  }
               }
            }
```

And use `list_controller@<stuff>` in `web.php`. Yeah, it's not as generic, but  the MigrationController is going away, and there's not likely to be more controllers (and if we do end up there, this could be changed to fix any snakecase contoller to the right versioned camelcase class).

Reply:
Then we are back to the discussion about the architecture of the whole thing.
As written above the original idea was to use the defined controller in web.php as a fallback controller if the middleware does resolve it differently.
And then there is the matter of, as you also write, that using an alias in the web.php structure is not really following the original concept of route definitions in Laravel/Lumen.

I have to give your idea an extra thought.

Reply:
>  As written above the original idea was to use the defined controller in web.php as a fallback controller if the middleware does resolve it differently.

Ah, I didn't realize you explicitly *wanted* the controller mentioned. That conflicts with my reasoning for *not* wanting it mentioned explicitly.

I can see the idea of having a fallback in `web.php` but I question its value. Sure, you can pull out the middleware and the service will still "work", but I think gives a false sense of security. Yes, it'll answer requsts to `/list`, but it'll give the *wrong* answer for all requests that asks for v2. So I consider the VersionSwitcher such an integral part of the service that I don't mind if things break if you pull it out. 

Actually, I'd prefer that things break if VersionSwitcher doesn't work. Imagine a small bug in the VersionSwitcher that makes it fail to find the right class if there's trailing whitespace in the header. If things just happily chug along with v1, nobody might notice at first (I'm sure you've run into developers that, faced with a disparity between doc and service output will just shrug and work with what the service returns and not ask upstream), until someone properly reads the documentation and points out that's not what the service over here gets as reply. Then you realize you can't actually fix the bug without breaking the service for live clients.

>  And then there is the matter of, as you also write, that using an alias in the web.php structure is not really following the original concept of route definitions in Laravel/Lumen.

It's stretching it a bit, for sure. But using the container for resolving the right implementation is right by the book. Usually it's done with interfaces, but ailases is designed for the same thing.